providers/implementations/serializers/deserialize_der2rsa.c

```diff
@@ -0,0 +1,231 @@
+                == NULL))
+            return 0;
+    }
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_DESERIALIZER_PARAM_PASS))
+        != NULL) {
+        OPENSSL_clear_free(ctx->sc.cipher_pass, ctx->sc.cipher_pass_length);
+        ctx->sc.cipher_pass = NULL;
+        if (!OSSL_PARAM_get_octet_string(p, &ctx->sc.cipher_pass, 0,
+                                         &ctx->sc.cipher_pass_length))
+            return 0;
+    }
+    return 1;
+}
+
+static int der2rsa_deserialize(void *vctx, OSSL_CORE_BIO *cin,
+                               OSSL_CALLBACK *data_cb, void *data_cbarg,
+                               OSSL_PASSPHRASE_CALLBACK *pw_cb, void *pw_cbarg)
+{
+    struct der2rsa_ctx_st *ctx = vctx;
+    void *libctx = PROV_LIBRARY_CONTEXT_OF(ctx->provctx);
+    RSA *rsa = NULL;
+    unsigned char *der = NULL;
+    const unsigned char *derp;
+    long der_len = 0;
+    unsigned char *new_der = NULL;
+    long new_der_len;
+    EVP_PKEY *pkey = NULL;       /* Oh, the irony */
+    int ok = 0;
+
+    ctx->sc.cb = pw_cb;
+    ctx->sc.cbarg = pw_cbarg;
+
+    if (!ossl_prov_read_der(ctx->provctx, cin, &der, &der_len))
+        return 0;
+
+    /*
+     * Opportunistic attempt to decrypt.  If it doesn't work, we try to
+     * decode our input unencrypted.
+     */
+    if (ctx->sc.cipher_intent
+        && ossl_prov_der_from_p8(&new_der, &new_der_len, der, der_len,
+                                 &ctx->sc)) {
+        OPENSSL_free(der);
+        der = new_der;
+        der_len = new_der_len;
+    }
+
+    derp = der;
+    if ((pkey = d2i_PrivateKey_ex(EVP_PKEY_RSA, NULL, &derp, der_len,
+                                  libctx, NULL)) != NULL) {
+        /* Tear out the RSA pointer from the pkey */
+        rsa = EVP_PKEY_get1_RSA(pkey);
+        EVP_PKEY_free(pkey);
+    }
+
+    OPENSSL_free(der);
+
+    if (rsa != NULL) {
+        OSSL_PARAM params[3];
+
+        params[0] =
+            OSSL_PARAM_construct_utf8_string(OSSL_DESERIALIZER_PARAM_DATA_TYPE,
+                                             "RSA", 0);
+        /* The address of the key becomes the octet string */
+        params[1] =
+            OSSL_PARAM_construct_octet_string(OSSL_DESERIALIZER_PARAM_REFERENCE,
+                                              &rsa, sizeof(rsa));
+        params[2] = OSSL_PARAM_construct_end();
+
+        ok = data_cb(params, data_cbarg);
+    }
+    RSA_free(rsa);
+
+    return ok;
+}
+
+static int der2rsa_export_object(void *vctx,
+                                 const void *reference, size_t reference_sz,
+                                 OSSL_CALLBACK *export_cb, void *export_cbarg)
+{
+    OSSL_FUNC_keymgmt_export_fn *rsa_export =
+        ossl_prov_get_keymgmt_rsa_export();
+    void *keydata;
+
+    if (reference_sz == sizeof(keydata) && rsa_export != NULL) {
+        /* The contents of the reference is the address to our object */
+        keydata = *(RSA **)reference;
```

Comment:
I'm not sure I understand why we pass around `RSA **` rather than just `RSA *`

Reply:
It's so that the `OSSL_FUNC_keymgmt_load()` function can "take over" object referred to, and signal that back to the deserializer function above by `*reference = NULL`.

The sequence of events is this:

1.  `der2rsa_deserialize` is called, and on success, it creates the OSSL_PARAM array, among others with the reference, which *in our case* is simply a pointer to a pointer to the RSA structure (so a pointer to `rsa`), and calls the data callback.
2.  The data callback, when determining that it can, calls `OSSL_FUNC_keymgmt_load()` with that same reference (i.e. it's contents is that pointer to a pointer to an RSA structure).
3.  `OSSL_FUNC_keymgmt_load()` grabs that reference, extracts an RSA object pointer from it, and then clears the original pointer (i.e. the `rsa` pointer from `der2rsa_deserialize()`).  In other words, it "steals" that object.

*This* function, however, doesn't need to steal anything, but must still extract the RSA object pointer the same way.

This is a hack, of course, but only applicable for our providers.  Other providers will have other kinds of references (PKCS#11 objects recently got some kind of opaque reference, I fully expect those to be passed around by a PKCS#11 provider).